lapce-ui/src/settings.rs

```diff
@@ -375,74 +376,56 @@ impl LapceSettings {
-                (
-                    "ui".to_string(),
-                    UIConfig::FIELDS.to_vec(),
-                    UIConfig::DESCS.to_vec(),
-                    settings,
-                )
-            }
-            LapceSettingsKind::Editor => {
-                let settings: HashMap<String, serde_json::Value> =
-                    serde_json::from_value(
-                        serde_json::to_value(&data.config.editor).unwrap(),
-                    )
-                    .unwrap();
-                (
-                    "editor".to_string(),
-                    EditorConfig::FIELDS.to_vec(),
-                    EditorConfig::DESCS.to_vec(),
-                    settings,
-                )
-            }
-            LapceSettingsKind::Terminal => {
-                let settings: HashMap<String, serde_json::Value> =
-                    serde_json::from_value(
-                        serde_json::to_value(&data.config.terminal).unwrap(),
-                    )
-                    .unwrap();
-                (
-                    "terminal".to_string(),
-                    TerminalConfig::FIELDS.to_vec(),
-                    TerminalConfig::DESCS.to_vec(),
-                    settings,
-                )
-            }
+        let (kind, fields, descs, mut settings) = match self.kind {
+            LapceSettingsKind::Core => (
+                "lapce",
+                &LapceConfig::FIELDS[..],
+                &LapceConfig::DESCS[..],
+                into_settings_map(&data.config.lapce),
+            ),
+            LapceSettingsKind::UI => (
+                "ui",
+                &UIConfig::FIELDS[..],
+                &UIConfig::DESCS[..],
+                into_settings_map(&data.config.ui),
+            ),
+            LapceSettingsKind::Editor => (
+                "editor",
+                &EditorConfig::FIELDS[..],
+                &EditorConfig::DESCS[..],
+                into_settings_map(&data.config.editor),
+            ),
+            LapceSettingsKind::Terminal => (
+                "terminal",
+                &TerminalConfig::FIELDS[..],
+                &TerminalConfig::DESCS[..],
+                into_settings_map(&data.config.terminal),
+            ),
         };
 
-        for (i, field) in fields.into_iter().enumerate() {
+        for (field, desc) in fields.iter().zip(descs.iter()) {
+            // TODO(dbuga): we should generate kebab-case field names
             let field = field.replace('_', "-");
+            let value = settings.remove(&field).unwrap();
             self.children.push(WidgetPod::new(
                 LapcePadding::new(
                     (10.0, 10.0),
                     LapceSettingsItem::new(
                         data,
-                        kind.clone(),
-                        field.clone(),
-                        descs[i].to_string(),
-                        settings.get(&field).unwrap().clone(),
+                        kind.to_string(),
```

Comment:
Ideally, all of these strings should be turned into `&'static str` but I did not want to limit flexibility. They might turn into `ReadOnlyString`s, though.